src/async.js

```diff
@@ -1,40 +1,56 @@
-// @ts-check
 import { WalkerBase } from './walker.js';
 
-/** @typedef { import('estree').BaseNode} BaseNode */
-/** @typedef { import('./walker').WalkerContext} WalkerContext */
-
-/** @typedef {(
+/**
+ * @typedef { import('estree').Node} Node
+ * @typedef { import('./walker.js').WalkerContext} WalkerContext
+ * @typedef {(
  *    this: WalkerContext,
- *    node: BaseNode,
- *    parent: BaseNode,
- *    key: string,
+ *    node: Node,
+ *    parent: Node,
+ *    key: string | number | symbol,
  *    index: number
- * ) => Promise<void>} AsyncHandler */
+ * ) => Promise<void>} AsyncHandler
+ */
 
 export class AsyncWalker extends WalkerBase {
 	/**
 	 *
-	 * @param {AsyncHandler} enter
-	 * @param {AsyncHandler} leave
+	 * @param {AsyncHandler} [enter]
+	 * @param {AsyncHandler} [leave]
 	 */
 	constructor(enter, leave) {
 		super();
 
-		/** @type {AsyncHandler} */
+		/** @type {boolean} */
+		this.should_skip = false;
+
+		/** @type {boolean} */
+		this.should_remove = false;
+
+		/** @type {Node | null} */
+		this.replacement = null;
+
+		/** @type {WalkerContext} */
+		this.context = {
+			skip: () => (this.should_skip = true),
+			remove: () => (this.should_remove = true),
+			replace: (node) => (this.replacement = node)
+		};
+
+		/** @type {AsyncHandler | undefined} */
```

Comment:
If you assign `this.whatever = whatever` in the constructor, TypeScript can infer its type, meaning this type annotation is redundant. There are multiple occurrences across multiple files. It’s not wrong, but personally I would omit them.